### PR TITLE
Enable reusuable_data memory mode by default

### PR DESCRIPTION
**Pull request by jack-berg**
This PR follows through on the plan to make reuseable data the default memory mode, as discussed in #6469.

I'm personally aware of it being used in several production systems for a while now without issue. If anyone is aware of any risks associated with this, please speak up! If not, this change in defaults will quietly improve the performance of all the opentelemetry-java users who missed the memory mode optimizations we've been making.

### CODE REVIEW

File: exporters/otlp/testing-internal/src/main/java/io/opentelemetry/exporter/otlp/testing/internal/FakeTelemetryUtil.java [hunk 1]
```diff
@@ -69,23 +74,15 @@ public static SpanData generateFakeSpanData() {
...
   /** Generate a fake {@link LogRecordData}. */
   public static LogRecordData generateFakeLogRecordData() {
     return TestLogRecordData.builder()
         .setResource(Resource.getDefault())
-        .setInstrumentationScopeInfo(
-            InstrumentationScopeInfo.builder("testLib")
-                .setVersion("1.0")
-                .setSchemaUrl("http://url")
-                .build())
+        .setInstrumentationScopeInfo(SCOPE_INFO)
```
Comment (jack-berg) [inline exporters/otlp/testing-internal/src/main/java/io/opentelemetry/exporter/otlp/testing/internal/FakeTelemetryUtil.java:85]: This change seems random, but is actually oddly needed for a subtle reason. The OTLP exporter tests generate fake test data, serialize and send it to a mock server, where it is deserialized into generated proto classes and evaluated for equality. As you're probably aware, OTLP payloads consist of a series of enclosing envelopes, e.g. for logs ResourceLogs enclose ScopeLogs enclose LogRecords. When an OTLP exporter recieves a `List<LogRecordData>`, the records need to be grouped by resource and scope to match the OTLP payload structure. We do this using _identity hash maps_. Notably, the reuseable memory mode serialization iterates over identity the identity hash maps slightly differently than the immutable memory mode variant. This means that when a group of records is exported with scopes which are equal but not identical, the order of those ScopeLogs in the serialized payload is not the same for immutable memory mode and reuseable memory mode.

We can make the tests pass by having the fake log records use identical scopes, which sidesteps the difference in iteration order between immutable memory mode and reuseable memory mode.

File: exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/PrometheusHttpServer.java [hunk 2]
```diff
@@ -82,7 +83,13 @@ public static PrometheusHttpServerBuilder builder() {
     // sequentially.
     if (memoryMode == MemoryMode.REUSABLE_DATA) {
       executor =
-          Executors.newSingleThreadExecutor(new DaemonThreadFactory("prometheus-http-server"));
```
Comment (jack-berg) [inline exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/PrometheusHttpServer.java:85]: `newSingleThreadExecutor` returns a delegating wrapped `ThreadPoolExecutor`. One of our tests introspects on the resolved executor and ensures that when memory mode is reuseable, a single thread executor is used. The test is not able to make the required assertions based on the delegating wrapped version returned by `Executors.newSingleThreadExecutor`.

File: exporters/otlp/testing-internal/src/main/java/io/opentelemetry/exporter/otlp/testing/internal/FakeTelemetryUtil.java (hunk 1 above)
Comment (breedx-splk) [reply exporters/otlp/testing-internal/src/main/java/io/opentelemetry/exporter/otlp/testing/internal/FakeTelemetryUtil.java:85]: Thanks for the clarification....this is a clever workaround.

File: exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/PrometheusHttpServer.java (hunk 2 above)
Comment (breedx-splk) [reply exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/PrometheusHttpServer.java:85]: The way you described it makes it sound like a shortcoming in the test/assertions...which is a bummer that the implementation had to change to accommodate it. It seems fine, but I can't fully reason about the impact of having a non-finalizable executorservice. I'm overthinking it.

File: exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/PrometheusHttpServerBuilder.java [hunk 3]
```diff
@@ -150,6 +151,11 @@ public PrometheusHttpServerBuilder setDefaultAggregationSelector(
    * registered with a {@link io.opentelemetry.sdk.metrics.SdkMeterProvider}.
    */
   public PrometheusHttpServer build() {
+    if (memoryMode == MemoryMode.REUSABLE_DATA && executor != null) {
+      throw new IllegalArgumentException(
+          "MemoryMode is REUSEABLE_DATA cannot be used with custom executor, "
+              + "since data may be corrupted if reading metrics concurrently");
```
Comment (breedx-splk) [inline exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/PrometheusHttpServerBuilder.java:157]: Is this documented anywhere else?

**Review by breedx-splk — APPROVED** [review]
Looking good to me. This is exciting!

File: exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/PrometheusHttpServerBuilder.java [hunk 4]
```diff
@@ -150,6 +151,11 @@ public PrometheusHttpServerBuilder setDefaultAggregationSelector(
    * registered with a {@link io.opentelemetry.sdk.metrics.SdkMeterProvider}.
    */
   public PrometheusHttpServer build() {
+    if (memoryMode == MemoryMode.REUSABLE_DATA && executor != null) {
+      throw new IllegalArgumentException(
+          "MemoryMode is REUSEABLE_DATA cannot be used with custom executor, "
```
Comment (jaydeluca) [inline exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/PrometheusHttpServerBuilder.java:156]:
```suggestion
          "MemoryMode is REUSEABLE_DATA and cannot be used with custom executor, "
```

File: exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/PrometheusHttpServerBuilder.java (hunk 3 above)
Comment (jaydeluca) [reply exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/PrometheusHttpServerBuilder.java:157]: we'll need to update the [docs](https://opentelemetry.io/docs/languages/java/configuration/) and change `otel.java.experimental.exporter.memory_mode` to `otel.java.exporter.memory_mode`, so we could potentially add in another footnote when we do that.

**Review by jaydeluca — APPROVED** [review]

**Pull request merged**